File: delightreading-server/src/main/java/com/delightreading/user/model/UserAuthenticationEntity.java
package com.delightreading.user.model;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.databind.PropertyNamingStrategy;
import com.fasterxml.jackson.databind.annotation.JsonNaming;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.Type;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;

import javax.persistence.*;
import java.io.Serializable;
import java.time.Instant;
import java.util.Collection;
import java.util.UUID;

import static javax.persistence.GenerationType.IDENTITY;

@Entity
@Table(name = "user_authentication")

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@JsonIgnoreProperties(ignoreUnknown = true)
@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonNaming(PropertyNamingStrategy.SnakeCaseStrategy.class)
public class UserAuthenticationEntity implements Serializable, UserDetails {


    ////////// Base {{
    @Id
    @GeneratedValue(strategy = IDENTITY)
    Long sid;

    @Column(name = "uid")
    String uid;

    @Column(name = "status")
    String status;

    @Column(name = "created_by")
    String createdBy;

    @Column(name = "created_at")
    Instant createdAt;

    @Column(name = "updated_by")
    String updatedBy;

    @Column(name = "updated_at")
    Instant updatedAt;


    @PrePersist
    public void prePersist() {
        if (uid == null) {
            uid = UUID.randomUUID().toString();
        }
        createdAt = Instant.now();
        // createdBy = LoggedUser.get();
    }

    @PreUpdate
    public void preUpdate() {
        updatedAt = Instant.now();
        // updatedBy = LoggedUser.get();
    }
    ////////// }} Base

    @ManyToOne(
            cascade = CascadeType.ALL
    )
    @JoinColumn(name="account_uid", referencedColumnName = "uid")
    UserAccountEntity account;

    @Column(name = "provider_id")
    String provider;

    @Column(name = "provider_account_id")
    String providerAccountId;

    @Column(name = "password")
    String password;

    @Column(name = "access_token")
    String accessToken;

    @Column(name = "refresh_token")
    String refreshToken;

    @Column(name = "expiration")
    Instant expiration;

    @Column(name = "raw_profile", length = 65535, columnDefinition="TEXT")
    @Type(type="text")
    String rawProfile;

    ///////// UserDetails {{
    @Override
    public Collection<? extends GrantedAuthority> getAuthorities() {
        return null;
    }

    @Override
    public String getUsername() {
        return this.getAccount().getUsername();
    }

    @Override
    public boolean isAccountNonExpired() {
        return (expiration == null || expiration.isBefore(Instant.now()));
    }

    @Override
    public boolean isAccountNonLocked() {
        return !UserAccountEntity.STATUS_LOCKED.equals(this.getAccount().getStatus());
    }

    @Override
    public boolean isCredentialsNonExpired() {
        return true;
    }

    @Override
    public boolean isEnabled() {
        return !UserAccountEntity.STATUS_DISABLED.equals(this.getAccount().getStatus());
    }
    ///////// }} UserDetails
}

File: delightreading-backend/src/main/kotlin/com/delightreading/user/model/UserAuthenticationEntity.kt
package com.delightreading.user.model

import org.hibernate.annotations.Type
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.userdetails.UserDetails

import java.io.Serializable
import java.time.Instant
import java.util.*
import javax.persistence.*

@Entity
@Table(name = "user_authentication")
data class UserAuthenticationEntity(

    //<editor-fold desc="Common Entity Base">
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var sid: Long? = null,

    @Column(name = "uid")
    var uid: String? = null,

    @Enumerated(EnumType.STRING)
    @Column(name = "status")
    var status: UserAccountStatus = UserAccountStatus.ACTIVE,

    @Column(name = "created_by")
    var createdBy: String? = null,

    @Column(name = "created_at")
    var createdAt: Instant = Instant.now(),

    @Column(name = "updated_by")
    var updatedBy: String? = null,

    @Column(name = "updated_at")
    var updatedAt: Instant? = null,
    //</editor-fold>

    @ManyToOne(cascade = arrayOf(CascadeType.ALL))
    @JoinColumn(name = "account_uid", referencedColumnName = "uid")
    var account: UserAccountEntity? = null,

    @Column(name = "provider_id")
    var provider: String? = null,

    @Column(name = "provider_account_id")
    var providerAccountId: String? = null,

    @Column(name = "password")
    var pwd: String? = null,

    @Column(name = "access_token")
    var accessToken: String? = null,

    @Column(name = "refresh_token")
    var refreshToken: String? = null,

    @Column(name = "expiration")
    var expiration: Instant? = null,

    @Column(name = "raw_profile", length = 65535, columnDefinition = "TEXT")
    @Type(type = "text")
    var rawProfile: String? = null

) : UserDetails, Serializable {

    //<editor-fold desc="Common Entity Base">
    @PrePersist
    fun prePersist() {
        if (uid == null) {
            uid = UUID.randomUUID().toString()
        }
        createdAt = Instant.now()
        // createdBy = LoggedUser.get();
    }

    @PreUpdate
    fun preUpdate() {
        updatedAt = Instant.now()
        // updatedBy = LoggedUser.get();
    }
    //</editor-fold>

    companion object {
        val LOCAL_PROVIDER = "local"
    }

    override fun getAuthorities(): Collection<GrantedAuthority>? {
        return null
    }

    override fun getUsername(): String? {
        return this.account?.username
    }

    override fun isAccountNonExpired(): Boolean {
        return expiration == null || expiration!!.isBefore(Instant.now())
    }

    override fun isAccountNonLocked(): Boolean {
        return this.account?.status != UserAccountStatus.LOCKED
    }

    override fun isCredentialsNonExpired(): Boolean {
        return true
    }

    override fun isEnabled(): Boolean {
        return this.account?.status != UserAccountStatus.DISABLED
    }

    override fun getPassword(): String? {
        return this.pwd
    }
}